zookeeper-api/src/main/java/org/apache/helix/zookeeper/impl/client/FederatedZkClient.java

```diff
@@ -80,19 +83,17 @@
   private PathBasedZkSerializer _pathBasedZkSerializer;
 
   // TODO: support capacity of ZkClient number in one FederatedZkClient and do garbage collection.
-  public FederatedZkClient(RealmAwareZkClient.RealmAwareZkClientConfig clientConfig,
-      MetadataStoreRoutingData metadataStoreRoutingData) {
-    if (metadataStoreRoutingData == null) {
-      throw new IllegalArgumentException("MetadataStoreRoutingData cannot be null!");
-    }
+  public FederatedZkClient(RealmAwareZkClient.RealmAwareZkClientConfig clientConfig)
+      throws IOException, InvalidRoutingDataException {
```

Comment:
I don't think it is a good idea to declare exception in a constructor which I would try to avoid. For this case, I would catch them and throw an `IllegalStateException`.

Reply:
Throwing exceptions in a constructor is not bad practice. In fact, it is the only way for a constructor to indicate that there is a problem; e.g. that the parameters/states are invalid.

This is a behavior agreed upon with @kaisun2000, which reminds me to add the Javadoc. IOException in the case of HTTP read failure, InvalidRoutingDataException in the case of an invalid routing data input. Users should be able to tell what's going on, and we let them distinguish failure cases by which type of exception we throw.

Reply:
I understand it is OK to throw exceptions in a constructor. In this case, if we declare exceptions in the signature, would it make it inconvenient to initialize/build a FederatedZkClient with try...catch? Personally, I would not like that. If the routing data does not exist or IO exception, it means routing data is not available for this FederatedZkClient construction. A runtime exception is good enough: `IllegalStateException("Routing data is not available: " + e.getMessage())`. I wouldn't bother try...catch to construct a FederatedZkClient.

Reply:
- FederatedZkClient is still a low-level construct
- This means that the users of this component will mostly be Helix Java APIs, and probably a few users who wish to use directly.
- In that case, I think we still need to float these exceptions up and handle each case at the Helix Java API level. That way, we make it 1) user-friendly and 2) we still achieve the low-level granularity with FederatedZkClient that users of this class would want instead of hiding failure types behind an IllegalStateException.

Reply:
Like I said, personally, hiding failure type in this case is more user-friendly than floating these exceptions up. And I also understand and respect your opinion.

 I am not sure how Helix Java APIs could handle these 2 exceptions differently when constructing this FederatedZkClient. Regarding a few other users, we could get opinions from them. I will be OK if we can get options from other users/committers and they think floating exceptions are better :)

Reply:
Here's one example where using checked exceptions is beneficial.

1) You could imagine a scenario where in the future we might have different sources from which we could obtain routing data. There is MSDS-1, MSDS-2, and MSDS-3 with different endpoints. In this case, we could add logic for trying options 1, 2, and 3 in order in case we run into IOException. Now, this logic won't apply to InvalidRoutingDataException because that implies that a data source is available but the actual routing data is invalid. This way, we could handle different exceptions differently.